### ISSUE TITLE
Database migration gets confused about foreign key columns that should live on gers

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1424]
_Imported from trac ticket 1424.
Created by **mpisarski**
Opened in trac: Jan 31, 2012, 3:34:12 PM
Last modified in trac: Feb 1, 2012, 2:17:05 PM_
***
_by mpisarski:_
```
CREATE TABLE creditapplications (
        applicationid INTEGER NOT NULL,
        accounttype VARCHAR(1 CHAR) NOT NULL,
        accountstatus VARCHAR(1 CHAR) NOT NULL,
        primaryapplicantid VARCHAR(255 CHAR) NOT NULL,
        secondaryapplicantid VARCHAR(255 CHAR),
        newsiteid VARCHAR(255 CHAR),
        createdby VARCHAR(255 CHAR),
        createdate DATE NOT NULL,
        modified DATE,
        modifiedby VARCHAR(255 CHAR),
        version INTEGER,
        PRIMARY KEY (applicationid),
        FOREIGN KEY(primaryapplicantid) REFERENCES creditapplicants (applicantid),
        FOREIGN KEY(secondaryapplicantid) REFERENCES creditapplicants (applicantid),
        FOREIGN KEY(modifiedby) REFERENCES users (username),
        FOREIGN KEY(createdby) REFERENCES "MISC"."EMP" ("EMP_CD")
)


2012-01-31 14:05:46,552 INFO sqlalchemy.engine.base.Engine {}
2012-01-31 14:05:46,556 INFO sqlalchemy.engine.base.Engine ROLLBACK
Traceback (most recent call last):
  File "/home/rarch/tg2env/bin/paster", line 8, in <module>
    load_entry_point('PasteScript==1.7.3', 'console_scripts', 'paster')()
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/command.py", line 84, in run
    invoke(command, command_name, options, args[1:])
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/command.py", line 123, in invoke
    exit_code = runner.run(args)
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/appinstall.py", line 68, in run
    return super(AbstractInstallCommand, self).run(new_args)
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/command.py", line 218, in run
    result = self.command()
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/appinstall.py", line 456, in command
    self, config_file, section, self.sysconfig_install_vars(installer))
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/appinstall.py", line 598, in setup_config
    mod.setup_app, command, filename, section, vars)
  File "/home/rarch/tg2env/lib/python2.6/site-packages/PasteScript-1.7.3-py2.6.egg/paste/script/appinstall.py", line 612, in _call_setup_app
    func(command, conf, vars)
  File "/home/rarch/trunk/src/appserver/pylotengine/websetup.py", line 66, in setup_app
    upgrade_database(model.metadata, gers_setup)
  File "/home/rarch/trunk/src/appserver/pylotengine/appsetup/upgrade/upgradedatabase.py", line 75, in upgrade_database
    gers_ddl = upgrade_tables(current_meta, new_meta, not gers_setup or gers_setup.create_gers_schema)
  File "/home/rarch/trunk/src/appserver/pylotengine/appsetup/upgrade/upgradedatabase.py", line 357, in upgrade_tables
    table.create()
  File "build/bdist.linux-x86_64/egg/sqlalchemy/schema.py", line 564, in create
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/base.py", line 2234, in _run_visitor
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/base.py", line 1904, in _run_visitor
  File "build/bdist.linux-x86_64/egg/sqlalchemy/sql/visitors.py", line 86, in traverse_single
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/ddl.py", line 86, in visit_table
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/base.py", line 1405, in execute
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/base.py", line 1496, in _execute_ddl
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/base.py", line 1646, in _execute_context
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/base.py", line 1639, in _execute_context
  File "build/bdist.linux-x86_64/egg/sqlalchemy/engine/default.py", line 330, in do_execute
sqlalchemy.exc.DatabaseError: (DatabaseError) ORA-02270: no matching unique or primary key for this column-list
 '\nCREATE TABLE creditapplications (\n\tapplicationid INTEGER NOT NULL, \n\taccounttype VARCHAR(1 CHAR) NOT NULL, \n\taccountstatus VARCHAR(1 CHAR) NOT NULL, \n\tprimaryapplicantid VARCHAR(255 CHAR) NOT NULL, \n\tsecondaryapplicantid VARCHAR(255 CHAR), \n\tnewsiteid VARCHAR(255 CHAR), \n\tcreatedby VARCHAR(255 CHAR), \n\tcreatedate DATE NOT NULL, \n\tmodified DATE, \n\tmodifiedby VARCHAR(255 CHAR), \n\tversion INTEGER, \n\tPRIMARY KEY (applicationid), \n\tFOREIGN KEY(primaryapplicantid) REFERENCES creditapplicants (applicantid), \n\tFOREIGN KEY(secondaryapplicantid) REFERENCES creditapplicants (applicantid), \n\tFOREIGN KEY(modifiedby) REFERENCES users (username), \n\tFOREIGN KEY(createdby) REFERENCES "MISC"."EMP" ("EMP_CD")\n)\n\n' {}
install.sh: failed trying to run setup-app
```[BR]
table definition in tables.py:
```
creditapplications_table = VersionedTable("creditapplications", metadata,
    Column("applicationid", Integer, primary_key=True),

    Column("accounttype", Unicode(1), nullable=False,
        info={'enum':'I,J,C'}),

    Column("accountstatus", Unicode(1), nullable=False,
        info={'enum':'N,E,P,S,D,A'}),

    Column("primaryapplicantid", Unicode(255), ForeignKey('creditapplicants.applicantid'), nullable=False,
        info={'required':False}),

    Column("secondaryapplicantid", Unicode(255), ForeignKey('creditapplicants.applicantid')),

    Column("newsiteid", Unicode(255)),

    Column("createdby", Unicode(255), ForeignKey('users.username')),

    Column("createdate", Date, nullable=False, default=datetime.now,
        info={'readonly':True}),

    Column("modified", Date),

    Column("modifiedby", Unicode(255), ForeignKey('users.username'))
)
```

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 31, 2012, 3:56:33 PM:_

Ticket #1423 has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Feb 1, 2012, 2:17:05 PM:_

(Changeset [[2262]](../commit/f59b3757ae85ef03bd23b247d4eff9b411559ac2))
 * fixes #1424 (Database migration gets confused about foreign key columns)

**Comment by kbower** [comment]
_**kbower** trac status change on Feb 1, 2012, 2:17:05 PM:_ **closed**
